File: backend/app/readiness/checkers/persistence.py
from __future__ import annotations

from app.api.errors import RuntimeBoundaryError
from app.infra.database import Database
from app.readiness.models import CheckStatus, CheckView


class PersistenceReadinessChecker:
    """检查数据库 Schema 兼容性和 SQLite WAL 持久化能力。"""

    name = "persistence"

    def __init__(self, database: Database) -> None:
        """绑定数据库；兼容基线由 Database 的唯一常量决定。"""
        self.database = database

    async def check(self) -> CheckView:
        """执行数据库版本与 journal mode 检查，不修改业务数据。"""
        try:
            schema = self.database.check_schema()
            journal_mode = self.database.journal_mode() if schema.writable else "blocked-read-only"
            details = {
                "journalMode": journal_mode,
                "schemaRevision": schema.revision,
                "persistentRoot": "configured",
            }
        except RuntimeBoundaryError as error:
            return CheckView(
                status=CheckStatus.BLOCKED,
                message=error.message,
                code=error.code,
                impact=error.impact,
                dataPreserved=error.data_preserved,
                schemaRevision=error.schema_revision,
                nextAction=error.next_action,
                details={"persistentRoot": "configured"},
            )
        if not schema.writable or journal_mode != "wal":
            return CheckView(
                status=CheckStatus.BLOCKED,
                message=schema.message if not schema.writable else "SQLite WAL 未启用",
                code=schema.code if not schema.writable else "PERSISTENCE_UNAVAILABLE",
                impact="业务数据无法安全持久化",
                dataPreserved=schema.data_preserved,
                schemaRevision=schema.revision,
                nextAction=schema.next_action or "检查 SQLite 数据库和持久化根目录",
                details=details,
            )
        return CheckView(
            status=CheckStatus.READY,
            message="持久化根目录和数据库可用",
            schemaRevision=schema.revision,
            details=details,
        )

File: backend/src/readiness/checkers/persistence.ts
import { RuntimeBoundaryError } from "../../api/errors.js";
import { Database } from "../../infra/database.js";
import { CheckView } from "../models.js";

/** 检查数据库 Schema 兼容性和 SQLite WAL 持久化能力。 */
export class PersistenceReadinessChecker {
  readonly name = "persistence";
  constructor(private readonly database: Database) {}
  /** 执行数据库版本与 journal mode 检查，不修改业务数据。 */
  async check(): Promise<CheckView> { try { const schema = this.database.checkSchema(); const details = { journalMode: schema.writable ? this.database.journalMode() : "blocked-read-only", schemaRevision: schema.revision, persistentRoot: "configured" }; return schema.writable ? { status: "ready", message: "持久化根目录和数据库可用", schemaRevision: schema.revision ?? undefined, details } : { status: "blocked", message: schema.message, code: schema.code ?? undefined, impact: "业务数据无法安全持久化", dataPreserved: schema.dataPreserved, schemaRevision: schema.revision ?? undefined, nextAction: schema.nextAction ?? "检查 SQLite 数据库和持久化根目录", details }; } catch (error) { const message = error instanceof RuntimeBoundaryError ? error.message : "持久化数据库当前不可用"; return { status: "blocked", message, code: error instanceof RuntimeBoundaryError ? error.code : "PERSISTENCE_UNAVAILABLE", impact: "业务数据无法安全持久化", dataPreserved: true, nextAction: "检查 SQLite 数据库和持久化根目录", details: { persistentRoot: "configured" } }; } }
}
